src/rgw/rgw_rados.cc

```diff
@@ -6306,77 +6306,460 @@ int RGWRados::Object::Read::read(int64_t ofs, int64_t end, bufferlist& bl, optio
-               uint64_t offset, optional_yield yield)
-    : store(store), client_cb(cb), aio(aio), offset(offset), yield(yield) {}
+  map<off_t, get_obj_io>::iterator liter = io_map.begin();
 
-  int flush(rgw::AioResultList&& results) {
-    int r = rgw::check_for_errors(results);
+  if (liter == io_map.end() ||
+      liter->first != ofs) {
+    return 0;
+  }
+
+  map<off_t, librados::AioCompletion *>::iterator aiter;
+  aiter = completion_map.find(ofs);
+  if (aiter == completion_map.end()) {
+    // completion map does not hold this io, it was cancelled
+    return 0;
+  }
+
+  AioCompletion *completion = aiter->second;
+  int r = completion->get_return_value();
+  if (r < 0)
+    return r;
+
+  for (; aiter != completion_map.end(); ++aiter) {
+    completion = aiter->second;
+    if (!completion->is_safe()) {
+      // reached a request that is not yet complete, stop
+        break;
+    }
+
+    r = completion->get_return_value();
     if (r < 0) {
+      set_cancelled(r); // mark it as cancelled, so that we don't continue processing next operations
       return r;
     }
 
-    auto cmp = [](const auto& lhs, const auto& rhs) { return lhs.id < rhs.id; };
-    results.sort(cmp); // merge() requires results to be sorted first
-    completed.merge(results, cmp); // merge results in sorted order
+    total_read += r;
 
-    while (!completed.empty() && completed.front().id == offset) {
-      auto bl = std::move(completed.front().data);
-      completed.pop_front_and_dispose(std::default_delete<rgw::AioResultEntry>{});
+    map<off_t, get_obj_io>::iterator old_liter = liter++;
+    bl_list.push_back(old_liter->second.bl);
+    io_map.erase(old_liter);
+  }
 
-      offset += bl.length();
-      int r = client_cb->handle_data(bl, 0, bl.length());
-      if (r < 0) {
-        return r;
-      }
-    }
-    return 0;
+  return 0;
+}
+
+static void _cache_aio_completion_cb(sigval_t sigval);
+
+std::vector<string> split(const string &s, const char * delim)
+{
+  stringstream ss(s);
+  string item;
+  std::vector<string> tokens;
+  while (getline(ss, item, (*delim))) {
+    tokens.push_back(item);
   }
+  return tokens;
+}
 
-  void cancel() {
-    // wait for all completions to drain and ignore the results
-    aio->drain();
+bool get_obj_data::deterministic_hash_is_local(string oid) {
+	return (deterministic_hash(oid).compare(cct->_conf->rgw_host)==0);
+}
+
+string get_obj_data::deterministic_hash(string oid)
+{
+  std::string location = cct->_conf->rgw_l2_hosts;
+  string delimiters(",");
+  std::vector<std::string> tokens = split(location, ",");
+  int mod = tokens.size();
+
+  std::string::size_type sz;   // alias of size_t
+  std::vector<string> sv = split(oid, "_");
+  /* Make sure the input string to stoi starts with a number */
+  std::string key = sv[sv.size() - 1];
+  std::string key_length = to_string(key.size());
+  int hash = std::stoi(key_length + key, &sz);
+  return tokens[hash%mod];
```

Comment:
`std::stoi()` throws exceptions that should probably be caught here

more generally, though, how is this a good hash function? what exactly is `key` here and why is the concatenation of `key_length + key` being parsed as an integer?

Reply:
is this the primary hash mechanism?  this doesn't sound like what D3N devs would have chosen...  If we lack a non-cryptographic hash, suggest XXHash;  if we require cryptographic, suggest blake2b

Reply:
this is part of the L2 distributed cache but was executing when it was disabled during the check if an object is local.
added a check to skip it if the distributed cache is explicitly disabled in the conf.

Reply:
> std::stoi() throws exceptions that should probably be caught here

adding try/catch

Reply:
this is related to the L2 separate PR, will address it there